mahidhar/segtree.cpp

```diff
@@ -0,0 +1,106 @@
+//This is a generic implementation of the segment tree without lazy propagation
+//This solves range sum query with updates
+//For other problems the implementation of the util class should be changed accordingly.
+
+#include <iostream>
+
+struct sTreeNode{
+	int x;
+};
+
+typedef struct sTreeNode snode;
+
+class util{
```

Comment:
As far as I understand, util class has a set of functions that create and return objects for struct. If my understanding is correct then this is not a good design because every time you return an object type the entire contents of return value will be copied to a temporary variable and then copied back to the actual object. A better method would be to create the object at the same position instead of using functions in util. Even better would be to overload constructors. Also, returning reference for the functions wouldn't be right idea because the variable `ans` in these functions will be destroyed once the function ends.

Reply:
Correct me if I missed understood something.

Reply:
the idea was to just change the functions in util class with touching the segtree class for other kind of problems. Also, I am not returning a reference if I understand it correctly. Also, can you please elaborate on this:
"A better method would be to create the object at the same position instead of using functions in util".
Thanks.